distributed/scheduler.py

```diff
@@ -8408,3 +8452,50 @@ def transition(
                 self.metadata[key] = ts.metadata
                 self.state[key] = finish
                 self.keys.discard(key)
+
+
+def coassignmnet_groups(
+    tasks: Sequence[TaskState], start: int = 0
+) -> dict[int, set[TaskState]]:
+    groups = {}
+    group = start
+    ix = 0
+    min_prio = None
+    max_prio = None
+
+    group_dependents_seen = set()
+    while ix < len(tasks):
+        current = tasks[ix]
+        if min_prio is None:
+            min_prio = ix
+
+        if current in group_dependents_seen or not current.dependents:
+            min_prio = None
+            max_prio = None
+            ix += 1
+            continue
+        # There is a way to implement this faster by just continuing to iterate
+        # over ix and check if the next is a dependent or not. I chose to go
+        # this route because this is what we wrote down initially
+        next = min(current.dependents, key=lambda ts: ts.priority)
+        next_ix = tasks.index(next)
```

Comment:
quadratic in `len(tasks)`.

Reply:
See comment about not optimising for performance. This is easily avoided when we want to do this properly but it was very easy and cheap to write this.

But still, why is this quadratic? I would expect a single pass to be sufficient to return an index 🤔 
are you flagging this because it's in another for loop, i.e. this line is what makes it quadratic. if so, you are absolutely right but as pointed out above we can actually rewrite this easily since the index increases strictly monotonically. I was just lazy when writing this

Reply:
The latter (inside another loop over tasks).